### PR TITLE
Add callback to lightning trainer

### PR DESCRIPTION
**Pull request by irasit**
## Description
Wire in the call back function for lightning estimator.

### CODE REVIEW

File: examples/pytorch/pytorch_lightning_mnist.py [hunk 1]
```diff
@@ -176,15 +175,27 @@ def test():
...
+            def on_init_start(self, trainer):
+                print('Starting to init trainer!')
+
+            def on_init_end(self, trainer):
+                print('trainer is init now')
+
+            def on_train_end(self, trainer, pl_module):
+                print('do something when training ends')
+
+        callbacks = [MyPrintingCallback(), ModelCheckpoint(dirpath=ckpt_path)]
```
Comment (chongxiaoc) [inline examples/pytorch/pytorch_lightning_mnist.py:190]: Is the `chkpt with bcast` hanging still there? What's the difference of the code now?

Comment (chongxiaoc) [reply examples/pytorch/pytorch_lightning_mnist.py:190]: Ignore above, I think that's for estimator training.

File: examples/spark/pytorch/pytorch_lightning_spark_mnist.py [hunk 2]
```diff
@@ -18,6 +18,7 @@
 from pyspark.sql.functions import udf
 
 from pytorch_lightning import LightningModule
+from pytorch_lightning.callbacks import EarlyStopping
```
Comment (chongxiaoc) [inline examples/spark/pytorch/pytorch_lightning_spark_mnist.py:21]: This is not used by now. Comment out?

File: examples/pytorch/pytorch_lightning_mnist.py (hunk 1 above)
Comment (irasit) [reply examples/pytorch/pytorch_lightning_mnist.py:190]: It works in this "pytorch_lightning_mnist.py", where estimator and petastorm loader is not used. 
It can be reproduced in "pytorch_lightning_spark_mnist.py " and the new test case that I added `test_model_checkpoint_callback` in this PR.

File: examples/spark/pytorch/pytorch_lightning_spark_mnist.py (hunk 2 above)
Comment (irasit) [reply examples/spark/pytorch/pytorch_lightning_spark_mnist.py:21]: Sure.

File: horovod/spark/lightning/estimator.py [hunk 3]
```diff
@@ -162,6 +162,10 @@ class TorchEstimator(HorovodEstimator, TorchEstimatorParamsWritable,
                                'Cache the data in memory for training and validation.',
                                typeConverter=TypeConverters.toBoolean)
 
+    num_gpus = Param(Params._dummy(), 'num_gpus',
+                     'number of gpus per process.')
```
Comment (chongxiaoc) [inline horovod/spark/lightning/estimator.py:166]: Can you explain and put a short description why we need to tell num_gpus per process for lightning?
I don't know how multi-gpus per single process/rank works for horovod.

File: horovod/spark/lightning/remote.py [hunk 4]
```diff
@@ -74,46 +75,65 @@ def RemoteTrainer(estimator, metadata, ckpt_bytes, run_id, dataset_idx, train_ro
...
-                'callbacks': callbacks,
-                'max_epochs': epochs,
-                'limit_train_batches': train_percent,
-                'limit_val_batches': val_percent,
-                'logger': logger,
-                'checkpoint_callback': checkpoint_callback,
-                'resume_from_checkpoint': (last_ckpt_file if ckpt_bytes else None),
-                'num_sanity_val_steps': 0
-            }
+                      'gpus': num_gpus,
```
Comment (chongxiaoc) [inline horovod/spark/lightning/remote.py:127]: It looks like lightning support num_gpus per process? How is gradients allreduce working in that case?

**Review by chongxiaoc — CHANGES_REQUESTED** [review]
I only have a question for num_gpus per process part.

File: horovod/spark/lightning/estimator.py (hunk 3 above)
Comment (irasit) [reply horovod/spark/lightning/estimator.py:166]: It is a parameter required by lightning trainer.

File: horovod/spark/lightning/remote.py (hunk 4 above)
Comment (irasit) [reply horovod/spark/lightning/remote.py:127]: It is defined [here](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/trainer.py#L201)
And used to get device id [here](https://github.com/PyTorchLightning/pytorch-lightning/blob/a584196abf820179adb0758ef67ddae91c44e7bc/pytorch_lightning/utilities/device_parser.py#L46)

Comment (chongxiaoc) [reply horovod/spark/lightning/remote.py:127]: This `num_gpus` means num_gpus per node.

Seeing this https://github.com/PyTorchLightning/pytorch-lightning/blob/a584196abf820179adb0758ef67ddae91c44e7bc/pytorch_lightning/trainer/connectors/accelerator_connector.py#L625

Horovod backend doesn't support `num_gpus` per node?

Comment (chongxiaoc) [reply horovod/spark/lightning/remote.py:127]: And this: https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/connectors/accelerator_connector.py#L599

It is saying no need to set n_gpus for horovod

File: examples/spark/pytorch/pytorch_lightning_spark_mnist.py [hunk 5]
```diff
@@ -124,13 +125,54 @@ def validation_epoch_end(self, outputs):
...
+            self.train_epcoh_end_counter += 1
+
+        def on_train_end(self, trainer, model):
+            print('Training ends')
+            assert self.epcoh_end_counter == 2 * epochs
+            assert self.train_epcoh_end_counter == epochs
+
+    callbacks = [MyDummyCallback()]
+
+    # FIXME: add EarlyStopping and ModelCheckpoint
```
Comment (tgaddair) [inline examples/spark/pytorch/pytorch_lightning_spark_mnist.py:165]: Maybe remove this for now? We don't want to confuse users who are just looking for an example of the API.

File: horovod/spark/lightning/estimator.py [hunk 6]
```diff
@@ -164,6 +164,10 @@ class TorchEstimator(HorovodEstimator, TorchEstimatorParamsWritable,
                                'Cache the data in memory for training and validation.',
                                typeConverter=TypeConverters.toBoolean)
 
+    num_gpus = Param(Params._dummy(), 'num_gpus',
```
Comment (tgaddair) [inline horovod/spark/lightning/estimator.py:167]: Should point out here that this will default to 1 when CUDA is available in the backend, otherwise 0.

File: horovod/spark/lightning/remote.py [hunk 7]
```diff
@@ -75,46 +76,65 @@ def RemoteTrainer(estimator, metadata, ckpt_bytes, run_id, dataset_idx, train_ro
     # Storage
     store = estimator.getStore()
     remote_store = store.to_remote(run_id, dataset_idx)
-    is_dbfs = isinstance(store, DBFSLocalStore)
+    #is_dbfs = isinstance(store, DBFSLocalStore)
```
Comment (tgaddair) [inline horovod/spark/lightning/remote.py:79]: If this is no longer needed, we should remove this line.

File: horovod/spark/lightning/remote.py [hunk 8]
```diff
@@ -75,46 +76,65 @@ def RemoteTrainer(estimator, metadata, ckpt_bytes, run_id, dataset_idx, train_ro
...
 
     val_steps_per_epoch = estimator.getValidationStepsPerEpoch()
     val_percent = val_rows / val_steps_per_epoch if val_steps_per_epoch else 1.0
 
-    # disable call back for now. Because petastorm can not reset index during training.
-    callbacks = None #_make_callbacks()
+    def train(serialized_model, num_gpus):
+        import horovod.torch as hvd
+        # Horovod: initialize library.
+        # hvd.init()
```
Comment (tgaddair) [inline horovod/spark/lightning/remote.py:90]: Why is this commented out?

File: examples/spark/pytorch/pytorch_lightning_spark_mnist.py (hunk 5 above)
Comment (irasit) [reply examples/spark/pytorch/pytorch_lightning_spark_mnist.py:165]: Sure, remove them for now.

File: horovod/spark/lightning/estimator.py (hunk 6 above)
Comment (irasit) [reply horovod/spark/lightning/estimator.py:167]: Sure, added.

File: horovod/spark/lightning/remote.py (hunk 7 above)
Comment (irasit) [reply horovod/spark/lightning/remote.py:79]: removed.

File: horovod/spark/lightning/remote.py (hunk 8 above)
Comment (irasit) [reply horovod/spark/lightning/remote.py:90]: Added back.

**Review by chongxiaoc — APPROVED** [review]
Peng you need a sign-off for commit

**Review by tgaddair — APPROVED** [review]
LGTM!
